File: selenium/src/test/java/com/ytfc/blog/pages/user/TestEmailPage.java
package com.ytfc.blog.pages.user;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.testng.Assert;

import com.ytfc.blog.utility.Helper;
import com.ytfc.blog.pages.PageObject;


public class TestEmailPage extends PageObject {
    public TestEmailPage(WebDriver driver) {
        this.driver = driver;
    }

    public TestEmailPage get(String type, String username) {
        Helper.get(driver, "test/view_email/" +type + "/" + username);
        return this; // Chainable
    }

    public String getValue(String field) {
        WebElement el = driver.findElement(By.id("email-" + field));
        return el.getText();
    }

    public String getGoodConfirmURL() {
        return getValue("urls-good-confirm");
    }

    public String getGoodDeclineURL() {
        return getValue("urls-good-decline");
    }

    // which = "id" or "key"
    public String getBadConfirmURL(String which) {
        return getValue("urls-bad-confirm-" + which);
    }

    public String getBadDeclineURL(String which) {
        return getValue("urls-bad-decline-" + which);
    }

    public String getGoodResetURL() {
        return getValue("urls-good-reset");
    }

    public String getBadResetURL(String which) {
        return getValue("urls-bad-reset-" + which);
    }

    public String getDataString(String id) {
        return getValue("data-" + id);
    }
    
    public void assertDataExists(String id, String message) {
        String actual = getDataString(id);
        Assert.assertTrue(actual.length() > 0, message);
    }

    public void assertData(String expected, String id, String message) {
        String actual = getDataString(id);
        Assert.assertEquals(actual, expected, message);
    }

    public int getDataInt(String id) {
        return Integer.parseInt(getValue("data-" + id));
    }
    
    public void assertData(int expected, String id, String message) {
        int actual = getDataInt(id);
        Assert.assertEquals(actual, expected, message);
    }

    public void assertValidEmail(String username, String message) {
        String expected = "Test email - " + username.toLowerCase() + "@example.com";
        String actual = getPageTitle();
        Assert.assertEquals(actual, expected, message);
    }

    public void assertNotValidEmail(String message) {
        Assert.assertEquals(getPageTitle(), "Test email not found", message);
    }

}

File: cypress/pages/test-email-page.js
import { Public } from './public'

export class TestEmailPage extends Public {
    constructor(type, username) {
        super();
        this._type = type
        this._name = username
    }

    getType() {
        return this._type
    }
    
    getName() {
        return this._name
    }
    
    pageUrl() {
        const type = this.getType()
        const name = this.getName()
        return `/test/view_email/${type}/${name}`;
    }

    confirmRegistration() {
        cy.get('td#email-urls-good-confirm > a').click();
        return this
    }

    declineRegistration() {
        cy.get('td#email-urls-good-decline > a').click();
        return this
    }

    badConfirmRegistration(reason) {
        cy.get(`td#email-urls-bad-confirm-${reason} > a`).click();
        return this
    }

    badDeclineRegistration(reason) {
        cy.get(`td#email-urls-bad-decline-${reason} > a`).click();
        return this
    }

    confirmReset() {
        cy.get('td#email-urls-good-reset > a').click();
        return this
    }

    badConfirmReset() {
        cy.get('td#email-urls-bad-reset-key > a').click();
        return this
    }

}
